### PR TITLE
Add "missing unsafe" diagnostics

### LINKED ISSUE #190: Expose unsafe operations
It would be great if the rust-analyzer would expose which operations are `unsafe`, e.g., calling an `unsafe fn`, dereferencing a pointer, creating a reference to a field of a packed struct, reading from an union, etc. So that Rust IDEs can underline/highlight these operations (e.g. because operations inside an unsafe block are not all always unsafe).

### PR DESCRIPTION
**Pull request by Nashenas88**
Addresses #190

### CODE REVIEW

File: crates/ra_hir/src/diagnostics.rs [hunk 1]
```diff
@@ -2,3 +2,53 @@
...
+use crate::db::HirDatabase;
+use crate::has_source::HasSource;
+use hir_ty::{
+    diagnostics::{MissingUnsafe, UnnecessaryUnsafe},
+    expr::unsafe_expressions,
+    InferenceResult,
+};
+use ra_syntax::AstPtr;
+
+pub struct UnsafeValidator<'a, 'b: 'a> {
```
Comment (Nashenas88) [inline crates/ra_hir/src/diagnostics.rs:18]: I added this validator here rather than in `ra_hir_ty` because I could not access `Function` in order to add the function as the source for the diagnostic messages. Is there a better location for this logic?

**Comment by Nashenas88** [comment]
Just realized I'm missing tests for the changes in ra_hir(_ty).

File: crates/ra_hir_ty/src/expr.rs [hunk 2]
```diff
@@ -310,3 +310,25 @@ pub fn record_pattern_missing_fields(
...
+        if let Expr::Call { callee, .. } = expr {
+            if infer
+                .method_resolution(*callee)
+                .map(|func| db.function_data(func).is_unsafe)
+                .unwrap_or(false)
+            {
+                unsafe_expr_ids.push(id);
+            }
+        }
+    }
```
Comment (bjorn3) [inline crates/ra_hir_ty/src/expr.rs:331]: This is missing raw pointer dereferences. Also taking references to packed structs should use unsafe in most cases. (currently always allowed without unsafe because of a not yet patched soundness hole)

Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs:344]: Just accounted for derefencing raw pointers. Regarding references to packed structs, is there a way to figure out which of the "most cases" are invalid?

File: crates/ra_hir_ty/src/expr.rs [hunk 3]
```diff
@@ -310,3 +315,72 @@ pub fn record_pattern_missing_fields(
...
+        let unsafe_expressions = unsafe_expressions(db, self.infer.as_ref(), def);
+        let func_data = db.function_data(self.func);
+        let unnecessary = func_data.is_unsafe && unsafe_expressions.len() == 0;
+        let missing = !func_data.is_unsafe && unsafe_expressions.len() > 0;
+        if !(unnecessary || missing) {
+            return;
+        }
+
+        let loc = self.func.lookup(db.upcast());
+        let in_file = loc.source(db.upcast());
```
Comment (Nashenas88) [inline crates/ra_hir_ty/src/expr.rs:375]: Should this be extracted into an fn somewhere? If so, where's the best place for something like that to live? Very similar logic exists in `ra_hir`, but due to concerns with cyclic dependencies, isn't available here.

File: crates/ra_hir_ty/src/expr.rs [hunk 4]
```diff
@@ -310,3 +315,87 @@ pub fn record_pattern_missing_fields(
...
+    infer: &InferenceResult,
+    def: DefWithBodyId,
+) -> Vec<ExprId> {
+    let mut unsafe_expr_ids = vec![];
+    let body = db.body(def);
+    for (id, expr) in body.exprs.iter() {
+        match expr {
+            Expr::Call { callee, .. } => {
+                if infer
+                    .method_resolution(/* id */ *callee)
```
Comment (Nashenas88) [inline crates/ra_hir_ty/src/expr.rs:330]: Can't seem to get this section to work. How do I get the `FunctionData` once I have an `Expr::Call` or `Expr::MethodCall`?

Comment (flodiebold) [reply crates/ra_hir_ty/src/expr.rs:330]: Look at the inferred type of `callee`. There's various callable types, but most of them can't be unsafe I think; for normal functions, the type will be `Ty::Apply(TypeCtor::FnDef(CallableDef::FunctionId(function_id)), ..)`. Fn pointers can also be unsafe, I think, but we don't track that yet.

File: crates/ra_hir_ty/src/expr.rs (hunk 2 above)
Comment (flodiebold) [reply crates/ra_hir_ty/src/expr.rs:359]: It's references to *fields of* packed structs, right? This seems like it might require some better tracking of place expressions, plus we'd need to keep track of struct representation in the first place.

File: crates/ra_hir_ty/src/expr.rs [hunk 5]
```diff
@@ -310,3 +315,87 @@ pub fn record_pattern_missing_fields(
...
+        sink: &'a mut DiagnosticSink<'b>,
+    ) -> UnsafeValidator<'a, 'b> {
+        UnsafeValidator { func, infer, sink }
+    }
+
+    pub fn validate_body(&mut self, db: &dyn HirDatabase) {
+        let def = self.func.into();
+        let unsafe_expressions = unsafe_expressions(db, self.infer.as_ref(), def);
+        let func_data = db.function_data(self.func);
+        let unnecessary = func_data.is_unsafe && unsafe_expressions.len() == 0;
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/expr.rs:383]: hmm, should we really lint about having `unsafe` on a function without unsafe operations? Especially since we won't detect all unsafe operations yet, so this will have lots of false positives, but even disregarding that, it can be correct for a function to be marked `unsafe` without it having any unsafe operations.

File: crates/ra_hir_ty/src/expr.rs (hunk 4 above)
Comment (flodiebold) [reply crates/ra_hir_ty/src/expr.rs:330]: For a method call, the code below should be correct.

File: crates/ra_ide/src/syntax_highlighting/tests.rs [hunk 6]
```diff
@@ -257,3 +257,33 @@ fn main() {
...
+unsafe fn unsafe_fn() {}
+
+struct HasUnsafeFn;
+
+impl HasUnsafeFn {
+    unsafe fn unsafe_method(&self) {}
+}
+
+fn main() {
+    let x = &5 as *usize;
```
Comment (bjorn3) [inline crates/ra_ide/src/syntax_highlighting/tests.rs:274]:
```suggestion
    let x = &5 as *const usize;
```

File: crates/ra_hir_ty/src/tests.rs [hunk 7]
```diff
@@ -532,6 +532,154 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
...
+#[test]
+fn no_missing_unsafe_diagnostic_with_unsafe_method_call_in_unsafe_block() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
+struct HasUnsafe;
+
+impl HasUnsafe {
+    unsafe fn unsafe_fn() {
+        let x = &5 as *usize;
```
Comment (bjorn3) [inline crates/ra_hir_ty/src/tests.rs:650]:
```suggestion
        let x = &5 as *const usize;
```

File: crates/ra_hir_ty/src/tests.rs [hunk 8]
```diff
@@ -532,6 +532,154 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
...
+    assert_snapshot!(diagnostics, @"");
+}
+
+#[test]
+fn no_missing_unsafe_diagnostic_with_unsafe_call_in_unsafe_block() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
+unsafe fn unsafe_fn() {
+    let x = &5 as *usize;
```
Comment (bjorn3) [inline crates/ra_hir_ty/src/tests.rs:624]:
```suggestion
    let x = &5 as *const usize;
```

File: crates/ra_hir_ty/src/tests.rs [hunk 9]
```diff
@@ -532,6 +532,154 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
...
+}
+
+#[test]
+fn no_missing_unsafe_diagnostic_with_raw_ptr_in_unsafe_block() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
+fn nothing_to_see_move_along() {
+    unsafe {
+        let x = &5 as *usize;
```
Comment (bjorn3) [inline crates/ra_hir_ty/src/tests.rs:606]:
```suggestion
        let x = &5 as *const usize;
```

File: crates/ra_hir_ty/src/tests.rs [hunk 10]
```diff
@@ -532,6 +532,154 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
...
+#[test]
+fn missing_unsafe_diagnostic_with_unsafe_method_call() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
+struct HasUnsafe;
+
+impl HasUnsafe {
+    unsafe fn unsafe_fn() {
+        let x = &5 as *usize;
```
Comment (bjorn3) [inline crates/ra_hir_ty/src/tests.rs:582]:
```suggestion
        let x = &5 as *const usize;
```

File: crates/ra_hir_ty/src/tests.rs [hunk 11]
```diff
@@ -532,6 +532,154 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
...
+    assert_snapshot!(diagnostics, @r#""fn missing_unsafe() {\n    let x = &5 as *usize;\n    let y = *x;\n}": Missing unsafe keyword on fn"#);
+}
+
+#[test]
+fn missing_unsafe_diagnostic_with_unsafe_call() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
+unsafe fn unsafe_fn() {
+    let x = &5 as *usize;
```
Comment (bjorn3) [inline crates/ra_hir_ty/src/tests.rs:558]:
```suggestion
    let x = &5 as *const usize;
```

File: crates/ra_hir_ty/src/tests.rs [hunk 12]
```diff
@@ -532,6 +532,154 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
     assert_snapshot!(diagnostics, @"");
 }
 
+#[test]
+fn missing_unsafe_diagnostic_with_raw_ptr() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
+fn missing_unsafe() {
+    let x = &5 as *usize;
```
Comment (bjorn3) [inline crates/ra_hir_ty/src/tests.rs:541]:
```suggestion
    let x = &5 as *const usize;
```

File: crates/ra_hir_ty/src/expr.rs (hunk 4 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs:330]: So I've been debugging this morning and the method call code doesn't work. `infer.method_resolution(id)` always returns `None`. Am I possibly setting up the inference incorrectly?

File: crates/ra_hir_ty/src/expr.rs (hunk 2 above)
Comment (bjorn3) [reply crates/ra_hir_ty/src/expr.rs:359]: > It's references to fields of packed structs, right?

Yes

File: crates/ra_hir_ty/src/expr.rs (hunk 5 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs]: I removed this diagnostic.

Comment (bjorn3) [reply crates/ra_hir_ty/src/expr.rs]: `Vec::set_len` is an example of an unsafe function that doesn't contain any unsafe operations.

File: crates/ra_hir_ty/src/tests.rs [hunk 13]
```diff
@@ -532,6 +532,140 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
...
+
+#[test]
+fn missing_unsafe_diagnostic_with_unsafe_method_call() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
+struct HasUnsafe;
+
+impl HasUnsafe {
+    unsafe fn unsafe_fn() {
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/tests.rs:581]: You're missing a `self` here, maybe that's the reason the test doesn't work?

Comment (Nashenas88) [reply crates/ra_hir_ty/src/tests.rs]: Thanks for that!

**Comment by Nashenas88** [comment]
I also realized there was no tracking in place for `unsafe` blocks. I've started adding support for that by initially just creating an `Expr::UnsafeBlock`, but I'm not sure how to determine whether an unsafe expr is within an unsafe block or not. Would I have to change the existing iteration in `unsafe_expressions` to walk the expression tree rather than iterating over every expr in body (currently done with `db.body(def).exprs.iter()`? I've been assuming the latter doesn't need to be manually walked.

File: crates/ra_hir_ty/src/expr.rs (hunk 2 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs:356]: Should I go ahead and track `Attrs` along with `StructData` in this work, or would it be easier to review as a follow-up addition?

Comment (flodiebold) [reply crates/ra_hir_ty/src/expr.rs:356]: I'm always for keeping PRs small ;)

**Comment by flodiebold** [comment]
> Expr::UnsafeBlock

That should probably just be an `Expr::Unsafe` with a block inside it.

> Would I have to change the existing iteration in unsafe_expressions to walk the expression tree rather than iterating over every expr in body (currently done with db.body(def).exprs.iter()? I've been assuming the latter doesn't need to be manually walked.

I think we'd need a function for walking the expression tree, we don't have that currently.

**Comment by Nashenas88** [comment]
@flodiebold You commented right after a figured out a working solution. Do you think what I came up with is not efficient? Should I still go ahead an work on a walking-based impl?

File: crates/ra_hir_ty/src/expr.rs (hunk 2 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs:384]: Created https://github.com/rust-analyzer/rust-analyzer/issues/4600 to track that separately. Cross referenced back here.

**Comment by Nashenas88** [comment]
I'm not understanding the test failures:
```
thread 'syntax_highlighting::tests::test_unsafe_highlighting' panicked at 'Forgot to remove debug-print?', crates/ra_ide/src/syntax_highlighting/tests.rs:288:5
```

But that points to this line:
```
assert_eq_text!(expected_html, actual_html);
```
in the new test I added. I can't reproduce this locally.

File: crates/ra_hir_ty/src/diagnostics.rs [hunk 14]
```diff
@@ -159,3 +162,31 @@ impl AstDiagnostic for BreakOutsideOfLoop {
         ast::Expr::cast(node).unwrap()
     }
 }
+
+#[derive(Debug)]
+pub struct MissingUnsafe {
+    pub file: HirFileId,
+    pub fn_def: AstPtr<ast::FnDef>,
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/diagnostics.rs:169]: I think this should point at the particular unsafe operation, not at the function. (and the message should go more in the direction of "This operation is unsafe and requires unsafe function or block", like in rustc)

File: crates/ra_hir_ty/src/expr.rs [hunk 15]
```diff
@@ -310,3 +315,112 @@ pub fn record_pattern_missing_fields(
...
+    def: DefWithBodyId,
+) -> Vec<UnsafeExpr> {
+    let mut unsafe_exprs = vec![];
+    let mut unsafe_block_scopes = vec![];
+    let body = db.body(def);
+    let expr_scopes = db.expr_scopes(def);
+    for (id, expr) in body.exprs.iter() {
+        match expr {
+            Expr::Unsafe { body } => {
+                if let Some(scope) = expr_scopes.scope_for(*body) {
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/expr.rs:342]: I don't know about using scopes for this. I don't think every block expression necessarily creates a new scope? As far as I can see, we only create new scopes when new names are introduced.

As an alternative solution that also doesn't involve walking, I could imagine creating a map from exprs to their parents in `Body` during lowering. You'd really just need to walk up and check whether there's an unsafe block somewhere in the expression's ancestors.

File: crates/ra_hir_ty/src/expr.rs [hunk 16]
```diff
@@ -310,3 +315,112 @@ pub fn record_pattern_missing_fields(
...
+                    unsafe_block_scopes.push(scope);
+                }
+            }
+            Expr::Call { callee, .. } => {
+                let ty = &infer.type_of_expr[*callee];
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/expr.rs:347]: you can use `infer[*callee]`

File: crates/ra_hir_ty/src/expr.rs [hunk 17]
```diff
@@ -310,3 +315,112 @@ pub fn record_pattern_missing_fields(
...
+                    if db.function_data(func).is_unsafe {
+                        unsafe_exprs.push(UnsafeExpr::new(id));
+                    }
+                }
+            }
+            Expr::MethodCall { .. } => {
+                if infer
+                    .method_resolution(id)
+                    .map(|func| db.function_data(func).is_unsafe)
+                    .unwrap_or_else(|| false)
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/expr.rs:362]: `unwrap_or(false)` is simpler (and not slower)

File: crates/ra_ide/src/syntax_highlighting.rs [hunk 18]
```diff
@@ -398,6 +398,18 @@ fn highlight_element(
...
         }
+        PREFIX_EXPR => {
+            let prefix_expr = element.into_node().and_then(ast::PrefixExpr::cast)?;
+            let expr = prefix_expr.expr()?;
+            let ty = sema.type_of_expr(&expr)?;
+            if !ty.is_raw_ptr() {
+                return None;
+            }
+
+            let mut h = Highlight::new(HighlightTag::Function);
```
Comment (flodiebold) [inline crates/ra_ide/src/syntax_highlighting.rs:409]: I don't know much about the highlighting code. Is `Function` the right tag to use here?

Also, shouldn't this check that this is a deref operation?

Comment (Nashenas88) [reply crates/ra_ide/src/syntax_highlighting.rs:409]: Oh, I forgot to follow-up on this. It seems I can't just add the unsafe modifier without some existing highlight tag. I used function initially just to see if I could get the ui tests working and forgot to circle back. Thanks for pointing this out. Who would be the best to bring in for this one? There's no tag for expressions, so I wasn't sure whether to add a new one or if one of the existing ones was meant to be used in cases like this.

File: crates/ra_hir_ty/src/expr.rs (hunk 17 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs:362]: Woops, I think this was left around from when I was trying to do something unnecessarily complicated. I'll clean it up.

File: crates/ra_hir_ty/src/diagnostics.rs (hunk 14 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/diagnostics.rs]: Done

File: crates/ra_hir_ty/src/expr.rs (hunk 16 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs]: Done

File: crates/ra_hir_ty/src/expr.rs (hunk 15 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs:340]: I'll follow up on this and the rest after work today.

File: crates/ra_hir_ty/src/expr.rs (hunk 3 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs]: No longer valid

File: crates/ra_hir_ty/src/expr.rs (hunk 15 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs]: Done

File: crates/ra_hir_def/src/body/lower.rs [hunk 19]
```diff
@@ -226,6 +265,7 @@ impl ExprCollector<'_> {
                         None => self.collect_expr_opt(condition.expr()),
                         // if let -- desugar to match
                         Some(pat) => {
+                            // FIXME(pfaria) track the break and arms parents here?
                             mark::hit!(infer_resolve_while_let);
                             let pat = self.collect_pat(pat);
                             let match_expr = self.collect_expr_opt(condition.expr());
```
Comment (Nashenas88) [inline crates/ra_hir_def/src/body/lower.rs:271]: When collecting for expr children within this section (up to the return below) should I include the inner ones created with `alloc_expr_desugared`?

File: crates/ra_ide/src/syntax_highlighting.rs (hunk 18 above)
Comment (Nashenas88) [reply crates/ra_ide/src/syntax_highlighting.rs:414]: Adding the case for the deref token. Still not sure who to contact about which tag to use or whether to create a new one.

File: crates/ra_hir_ty/src/expr.rs (hunk 15 above)
Comment (flodiebold) [reply crates/ra_hir_ty/src/expr.rs]: > should expressions have parents? I don't really have a great answer here, but my gut feeling is that an extra walk is probably better than having parent links, in this case

Hmm, expressions have an identity, to which e.g. their inferred type is tied (so it would be impossible for the same expression to have multiple parents) -- so they have parents either way, the question is just whether there's an efficient way to get them, in my opinion.

Comment (matklad) [reply crates/ra_hir_ty/src/expr.rs]: That's a very good point about identity, yeah. I think we should add parents if that's convenient then!

File: crates/ra_ide/src/syntax_highlighting.rs (hunk 18 above)
Comment (Nashenas88) [reply crates/ra_ide/src/syntax_highlighting.rs]: Added `HighlightTag::Operator`.

File: crates/ra_hir_ty/src/expr.rs [hunk 20]
```diff
@@ -310,3 +313,111 @@ pub fn record_pattern_missing_fields(
...
+                unsafe_expr.inside_unsafe_block = true;
+                continue 'unsafe_exprs;
+            }
+        }
+    }
+
+    unsafe_exprs
+}
+
+pub struct UnsafeValidator<'a, 'b: 'a> {
```
[the remark this answers is not in the document]
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs]: Done

File: crates/ra_hir_def/src/body/lower.rs (hunk 19 above)
Comment (Nashenas88) [reply crates/ra_hir_def/src/body/lower.rs:271]: @matklad or @flodiebold any idea what I should do in this section?

File: crates/ra_hir_ty/src/expr.rs [hunk 21]
```diff
@@ -310,3 +311,71 @@ pub fn record_pattern_missing_fields(
...
+    pub inside_unsafe_block: bool,
+}
+
+impl UnsafeExpr {
+    fn new(expr: ExprId) -> Self {
+        Self { expr, inside_unsafe_block: false }
+    }
+}
+
+pub fn unsafe_expressions(
```
Comment (matklad) [inline crates/ra_hir_ty/src/expr.rs:326]: I feel this function should be in the same module as `UnsafeValidator`.

File: crates/ra_hir_def/src/body/lower.rs [hunk 22]
```diff
@@ -160,11 +162,28 @@ impl ExprCollector<'_> {
...
+
+        parent_expr
+    }
+
     fn collect_expr(&mut self, expr: ast::Expr) -> ExprId {
+        let parent_and_children = self.collect_expr_inner(expr);
+        self.update_parent_map(parent_and_children)
+    }
+
+    fn collect_expr_inner(&mut self, expr: ast::Expr) -> (ExprId, Vec<ExprId>) {
```
[the remark this answers is not in the document]
Comment (Nashenas88) [reply crates/ra_hir_def/src/body/lower.rs:181]: I'm not sure I follow. The parents aren't constructed until after the children have already been constructed. So as your descending, you don't have the parent id. If we're just trying to avoid the vec allocation, I can do the mapping inline.

File: crates/ra_hir_def/src/body/lower.rs (hunk 19 above)
Comment (Nashenas88) [reply crates/ra_hir_def/src/body/lower.rs:271]: That `debug_assert` was a good idea! I've got most everything else covered now except for one test that panics because of adding that. Can't finish tonight, but will try pushing something out tomorrow morning.

File: crates/ra_hir_ty/src/expr.rs (hunk 21 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/expr.rs]: Done

File: crates/ra_hir_def/src/body/lower.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (Nashenas88) [reply crates/ra_hir_def/src/body/lower.rs]: Let me know what you think about the macro version. I think this avoids adding anything extra to `ExprCollector` and also avoids allocing too many extra `Vec`s to pass back. I couldn't avoid cloning existing vecs though. Since I can't get the parent expr id until I've already passed the needed children expr id vecs to the constructing call.

File: crates/ra_hir_def/src/body.rs [hunk 23]
```diff
@@ -184,6 +184,7 @@ pub struct Body {
     /// The `ExprId` of the actual body expression.
     pub body_expr: ExprId,
     pub item_scope: ItemScope,
+    pub parent_map: FxHashMap<ExprId, ExprId>,
```
[the remark this answers is not in the document]
Comment (Nashenas88) [reply crates/ra_hir_def/src/body.rs]: Done

File: crates/ra_hir_def/src/body/lower.rs (hunk 19 above)
Comment (Nashenas88) [reply crates/ra_hir_def/src/body/lower.rs]: Is the current `debug_assert` sufficient? I'm checking that all children found during walk exist in the map and that the counts exactly match how many children map to parents. Should I add something that logs which child ids are extra/missing?

File: crates/ra_hir_def/src/body/lower.rs [hunk 24]
```diff
@@ -84,6 +85,40 @@ struct ExprCollector<'a> {
...
+            track_parent!(@build $collector, $parent $($rest)*)
+        }
+    };
+    ($collector:ident, $parent:expr, $($rest:tt)*) => {
+        {
+            let parent = $parent;
+            track_parent!(@build $collector, parent, $($rest)*)
+        }
+    }
+}
```
Comment (matklad) [inline crates/ra_hir_def/src/body/lower.rs:120]: In general, we use macros sparingly in rust-analyzer. For "just remove some copy paste" there's almost always a better solution, which typically is:

* just write a bit of boilerplate
* write some type-level or runtime code to do this. 

In this case, I think what would work is some runtime tracking. For example, we could allocate a single `Vec` with children or something like that. i am not entirely sure about this, but probability that the macro is the only non-allocating solution seems pretty low to me. 

Additionally, in this case I think there's an obvious easy solution -- let's remove parent tracking altogether, and instead turn an `assert` about parents into just the code that computes the parents mapping.

File: crates/ra_hir_ty/src/unsafe_validation.rs [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+use hir_expand::diagnostics::DiagnosticSink;
+
+use crate::{
+    db::HirDatabase, diagnostics::MissingUnsafe, lower::CallableDef, ApplicationTy,
+    InferenceResult, Ty, TypeCtor,
+};
+
+use rustc_hash::FxHashSet;
+
+pub use hir_def::{
```
Comment (matklad) [inline crates/ra_hir_ty/src/unsafe_validation.rs:16]: Seems like this shouldn't be `pub`?

File: crates/ra_ide/src/syntax_highlighting.rs [hunk 26]
```diff
@@ -450,7 +467,13 @@ fn highlight_name(db: &RootDatabase, def: Definition) -> Highlight {
...
         Definition::ModuleDef(def) => match def {
             hir::ModuleDef::Module(_) => HighlightTag::Module,
-            hir::ModuleDef::Function(_) => HighlightTag::Function,
+            hir::ModuleDef::Function(func) => {
+                let mut h = HighlightTag::Function.into();
+                if func.is_unsafe(db) {
+                    h |= HighlightModifier::Unsafe;
+                }
+                return h;
+            }
```
Comment (matklad) [inline crates/ra_ide/src/syntax_highlighting.rs:476]: Wait, so, for actual unsafe highlighting, we don't even need to track unsafe context, we only need types? 

So, we can implement the highlighting bits without changes to lowering and such? Let's split this off into a separate PR then!

Comment (Nashenas88) [reply crates/ra_ide/src/syntax_highlighting.rs:476]: that makes sense. I'll split that off after work tonight. Should I also implement highlighting within vs code? Something like https://github.com/rust-analyzer/rust-analyzer/pull/4603/files#diff-9234de0bfa99228a67ea76a4391c86f1R598

Comment (matklad) [reply crates/ra_ide/src/syntax_highlighting.rs:476]: If you add news tags/modifiers, they should be specified in package.json. But it seems like we already have operator and unsafe, so that should not be necessary

File: crates/ra_hir_def/src/body/lower.rs (hunk 24 above)
Comment (Nashenas88) [reply crates/ra_hir_def/src/body/lower.rs]: Thanks to your persuasion, I came up with something much smaller. It takes advantage of `walk_child_exprs` and the only downside is it walks all immediate child exprs of an expr even if it already found an answer (`walk_child_exprs` doesn't support early return).

File: crates/ra_hir_ty/src/unsafe_validation.rs (hunk 25 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/unsafe_validation.rs]: Fixed

File: crates/ra_ide/src/syntax_highlighting.rs (hunk 26 above)
Comment (Nashenas88) [reply crates/ra_ide/src/syntax_highlighting.rs]: Opened #4720

File: crates/ra_hir_ty/src/unsafe_validation.rs [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+    }
+}
+
+pub fn unsafe_expressions(
+    db: &dyn HirDatabase,
+    infer: &InferenceResult,
+    def: DefWithBodyId,
+) -> Vec<UnsafeExpr> {
+    let mut unsafe_exprs = vec![];
+    let mut unsafe_block_exprs = FxHashSet::default();
```
Comment (matklad) [inline crates/ra_hir_ty/src/unsafe_validation.rs:78]: Seems like this is unused?

File: crates/ra_hir_ty/src/unsafe_validation.rs [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+use hir_expand::diagnostics::DiagnosticSink;
+
+use crate::{
+    db::HirDatabase, diagnostics::MissingUnsafe, lower::CallableDef, ApplicationTy,
+    InferenceResult, Ty, TypeCtor,
+};
+
+use rustc_hash::FxHashSet;
+
+use hir_def::{
```
Comment (matklad) [inline crates/ra_hir_ty/src/unsafe_validation.rs:16]: https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/README.md#order-of-imports

File: crates/ra_hir_ty/src/unsafe_validation.rs [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+    body::Body,
+    expr::{Expr, ExprId, UnaryOp},
+};
+
+pub struct UnsafeValidator<'a, 'b: 'a> {
```
Comment (matklad) [inline crates/ra_hir_ty/src/unsafe_validation.rs:21]: Seems like this (and other stuff) can be `pub(crate)`?

File: crates/ra_hir_ty/src/unsafe_validation.rs [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+    if let &Expr::Unsafe { body: child } = expr {
+        return is_in_unsafe(body, child, needle, true);
+    }
+
+    let mut found = false;
+    expr.walk_child_exprs(|child| {
+        found = found || is_in_unsafe(body, child, needle, within_unsafe);
+    });
+    found
+}
```
Comment (matklad) [inline crates/ra_hir_ty/src/unsafe_validation.rs:138]: Yup, this seems like much more targeted and maintainable approach! This does asymptotically unfortunate number of traversals, but this should be find in practice (there are few unsafe exps) and should be easy to fix locally by adding some kind of mapping.

File: crates/ra_hir_ty/src/tests.rs [hunk 31]
```diff
@@ -532,6 +532,160 @@ fn missing_record_pat_field_no_diagnostic_if_not_exhaustive() {
...
+    .0;
+
+    assert_snapshot!(diagnostics, @r#""HasUnsafe.unsafe_fn()": This operation is unsafe and requires an unsafe function or block"#);
+}
+
+#[test]
+fn no_missing_unsafe_diagnostic_with_raw_ptr_in_unsafe_block() {
+    let diagnostics = TestDB::with_files(
+        r"
+//- /lib.rs
```
Comment (matklad) [inline crates/ra_hir_ty/src/tests.rs:603]: No need to specify file name for single file.

File: crates/ra_hir_ty/src/unsafe_validation.rs (hunk 29 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/unsafe_validation.rs:21]: These are all used in `rust_hir`, so making them `pub(crate)` would break that crate. This is following the same structure as `ra_hir_ty::expr::ExprValidator`.

File: crates/ra_hir_ty/src/tests.rs (hunk 31 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/tests.rs]: Done

File: crates/ra_hir_ty/src/unsafe_validation.rs (hunk 28 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/unsafe_validation.rs]: Done

File: crates/ra_hir_ty/src/unsafe_validation.rs (hunk 27 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/unsafe_validation.rs]: Removed

File: crates/ra_hir_ty/src/unsafe_validation.rs (hunk 30 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/unsafe_validation.rs:121]: Does the latest commit seem better? I'm doing the walk and collection at once now, so it should just visit every expr in the body once.

**Comment by matklad** [comment]
bors d+
